chi/widgets.py

```diff
@@ -175,125 +168,88 @@ def update_selected_node(node_type):
         with node_output:
             chi.use_node(node_type, avail_nodes[node_type], verbose)
 
-    if get_nodes(display=False) is None:
-        return
-    avail_nodes = get_nodes(display=False)[0]
+    avail_nodes = get_nodes(display=False)["avail"]
 
-    def _find_gpu_helper(nodes, req_count: int = None):
-        """Find all nodes with or without GPUs, and restrict the set to nodes
-        with a certain number of GPUs if specified.
-        """
+    def has_gpu_helper(nodes, req_gpu):
+        """Find all nodes with or without GPUs."""
         new_nodes = {}
         for node_type, data in nodes.items():
-            if req_count is None:
-                if "gpu" in data and data["gpu"]["gpu"] is True:
-                    new_nodes[node_type] = data
-            elif req_count == 0 and ("gpu" not in data or
-                                     data["gpu"]["gpu"] is False):
+            gpu_exists = not ("gpu" not in data or data["gpu"]["gpu"] is False)
+            if gpu_exists and req_gpu:
                 new_nodes[node_type] = data
-            elif req_count > 0 and "gpu" in data and data["gpu"]["gpu"] is \
-                    True:
-                true_count = data["gpu"]["gpu_count"]
-                if true_count == req_count:
-                    new_nodes[node_type] = data
+            if not gpu_exists and not req_gpu:
+                new_nodes[node_type] = data
+        return new_nodes
+
+    def gpu_count_helper(nodes, req_count):
+        """Find all nodes with exactly REQ_COUNT GPUs."""
+        new_nodes = {}
+        for node_type, data in nodes.items():
+            gpu_exists = not ("gpu" not in data or data["gpu"]["gpu"] is False)
+            if gpu_count < 0:
```

Comment:
Where does `gpu_count` come from?

Reply:
I see now. This is a very large function with lots of nested functions. We do admittedly have a lot of code like this, but it's not great for readability.